### PR TITLE
Add Bors merge bot

### PR DESCRIPTION
**Pull request by smnzhu**
This PR adds the required configuration files for setting up [bors](https://bors.tech/) mergebot.

Configuration and usage guide: https://bors.tech/documentation/

Important commands:
* When ready to merge a PR, comment on the PR with the command "bors merge"
* Cancel a merge: "bors cancel"

Here is a good [explanation](https://github.com/bors-ng/bors-ng#:~:text=As%20commits%20are%20reviewed%2C%20bors%20lumps%20them%20into%20a%20queue%20of%20batches.%20If%20everything%20passes%2C%20there%20will%20just%20be%20two%20batches%3B%20the%20one%20that%27s%20running%2C%20and%20the%20one%20that%27s%20waiting%20to%20be%20run%20(and%20is%20accumulating%20more%20and%20more%20pull%20requests%20until%20it%20gets%20a%20chance%20to%20run).) of `bors`'s merge strategy.

Bors will create two new branches `staging` and `trying`. [Here](https://bors.tech/documentation/getting-started/#:~:text=Once%20you%E2%80%99ve,be%20well-isolated.) is the explanation of what they are used for.

## TODO:
- [x] Install bors app on flow-go repo: https://github.com/apps/bors/installations/new
- [x] Connect bors within Github: https://github.com/apps/bors
- [ ] Update branch protection settings to add bors as a merge-blocking status check for PRs to `master`, and remove existing CI jobs as merge-blocking
- [ ] Update onboarding notion docs / instructions with bors usage instructions

### CODE REVIEW

File: .github/workflows/ci.yml
Comment (smnzhu) [inline .github/workflows/ci.yml:6]: We can remove this line imo, since bors will be the only one pushing to master (we should disable pushing to master in the repo settings).

File: .github/workflows/ci.yml
Comment (smnzhu) [inline .github/workflows/ci.yml:12]: We can optionally remove this as well, since status checks will already be run by bors before merging. 

We can keep it if we still want to run all of them automatically on every PR, even though they will no longer be merge blocking.

File: bors.toml
Comment (smnzhu) [inline bors.toml:10]: https://bors.tech/documentation/#:~:text=Configuration%20(bors.toml)

File: .github/workflows/ci.yml
Comment (huitseeker) [reply .github/workflows/ci.yml:6]: That's correct in the general case, but I think we are going to need a transition phase where merges to master occur both through PRs and manually, hence this remains useful.

File: .github/workflows/ci.yml
Comment (huitseeker) [reply .github/workflows/ci.yml:12]: The payload of what bors tests and the one that CI runs are different. If I open a PR "foo" on top of master1 in July, and it gets approved in August, by which time master has advanced to master2.

CI tests master1 + foo
Bors tests master2 + foo

Both are useful to signal me (the PR author) if something is problematic, at different stages.

**Review by huitseeker — COMMENTED** [review]
One important thing I think we should adopt for bors to deal with flakiness is to use a job that re-runs CI several times, such as https://github.com/onflow/flow-go/pull/1004, because of the ongoing flakiness. This should incur no changes to the *normal* CI, which should give more interactive feedback to developers. WDYT?

Another item we may want to sync on: do we want to run the publicly hosted variant of Bors, or do we want to host ours?

**Comment by smnzhu** [comment]
> One important thing I think we should adopt for bors to deal with flakiness is to use a job that re-runs CI several times, such as https://github.com/onflow/flow-go/pull/1004, because of the ongoing flakiness. This should incur no changes to the *normal* CI, which should give more interactive feedback to developers. WDYT?
> 
> Another item we may want to sync on: do we want to run the publicly hosted variant of Bors, or do we want to host ours?

I definitely agree with the first point, and was actually about to mention this. 

I like the idea of having a separate job for this.

Regarding the second point, what would the difference be?

**Comment by smnzhu** [comment]
> One important thing I think we should adopt for bors to deal with flakiness is to use a job that re-runs CI several times, such as #1004, because of the ongoing flakiness. This should incur no changes to the _normal_ CI, which should give more interactive feedback to developers. WDYT?

Done: https://github.com/onflow/flow-go/pull/1066/files/65f33baa52b3ad36651f0988addece03b5ff4178..ca8963a963e6cfe617a80d9b9753c99335804605

File: bors.toml
Comment (huitseeker) [inline bors.toml:6]: Nit: we may be able to simplify this using [SQL LIKE](https://www.w3schools.com/sql/sql_like.asp) syntax

**Review by huitseeker — COMMENTED** [review]
Approval conditional on solving TODOs & hosting issue

> Regarding the second point, what would the difference be?

A third-party application has push rights to master.

File: .github/workflows/ci.yml
Comment (Kay-Zee) [inline .github/workflows/ci.yml:9]: are we removing these later?

File: CONTRIBUTING.md
Comment (Kay-Zee) [inline CONTRIBUTING.md:116]: so for now the only thing that changes is instead of pressing the big green button, we comment `bors merge`? Just to double check my understanding is correct.

**Review by Kay-Zee — APPROVED** [review]

File: .github/workflows/ci.yml
Comment (smnzhu) [reply .github/workflows/ci.yml:9]: No, these are needed by bors. It pushes to these branches to run CI. See [this](https://bors.tech/documentation/getting-started/#:~:text=The%20pull%20request%2C%20as,not%20be%20well-isolated.)

File: CONTRIBUTING.md
Comment (smnzhu) [reply CONTRIBUTING.md:116]: Yep, that's correct.

**Review by huitseeker — APPROVED** [review]
🎉

**Comment by smnzhu** [comment]
bors merge

**Pull request merged**
